lms/static/scripts/frontend_apps/components/test/EmailNotificationsPreferences-test.js

```diff
@@ -0,0 +1,103 @@
+    'instructor_email_digests.days.mon': true,
+    'instructor_email_digests.days.tue': false,
+    'instructor_email_digests.days.wed': true,
+    'instructor_email_digests.days.thu': false,
+    'instructor_email_digests.days.fri': false,
+    'instructor_email_digests.days.sat': true,
+  };
+
+  beforeEach(() => {
+    fakeSetSelectedDays = sinon.stub();
+  });
+
+  function createComponent() {
+    return mount(
+      <EmailNotificationsPreferences
+        selectedDays={initialSelectedDays}
+        setSelectedDays={fakeSetSelectedDays}
+      />
+    );
+  }
+
+  function getCheckbox(wrapper, day) {
+    return wrapper.find(
+      `Checkbox[data-testid="instructor_email_digests.days.${day}-checkbox"]`
+    );
+  }
+
+  it('initial selects appropriate checkboxes', () => {
+    const wrapper = createComponent();
+
+    assert.isTrue(getCheckbox(wrapper, 'sun').prop('checked'));
+    assert.isTrue(getCheckbox(wrapper, 'mon').prop('checked'));
+    assert.isFalse(getCheckbox(wrapper, 'tue').prop('checked'));
+    assert.isTrue(getCheckbox(wrapper, 'wed').prop('checked'));
+    assert.isFalse(getCheckbox(wrapper, 'thu').prop('checked'));
+    assert.isFalse(getCheckbox(wrapper, 'fri').prop('checked'));
+    assert.isTrue(getCheckbox(wrapper, 'sat').prop('checked'));
+  });
+
+  it('can select all checkboxes', () => {
+    const wrapper = createComponent();
+
+    wrapper.find('button[data-testid="select-all-button"]').simulate('click');
+
+    assert.calledWith(fakeSetSelectedDays, {
+      'instructor_email_digests.days.sun': true,
+      'instructor_email_digests.days.mon': true,
+      'instructor_email_digests.days.tue': true,
+      'instructor_email_digests.days.wed': true,
+      'instructor_email_digests.days.thu': true,
+      'instructor_email_digests.days.fri': true,
+      'instructor_email_digests.days.sat': true,
+    });
+  });
+
+  it('can unselect all checkboxes', () => {
+    const wrapper = createComponent();
+
+    wrapper.find('button[data-testid="select-none-button"]').simulate('click');
+
+    assert.calledWith(fakeSetSelectedDays, {
+      'instructor_email_digests.days.sun': false,
+      'instructor_email_digests.days.mon': false,
+      'instructor_email_digests.days.tue': false,
+      'instructor_email_digests.days.wed': false,
+      'instructor_email_digests.days.thu': false,
+      'instructor_email_digests.days.fri': false,
+      'instructor_email_digests.days.sat': false,
+    });
+  });
+
+  ['sun', 'mon', 'thu'].forEach(day => {
+    it('lets individual days to be changed', () => {
+      const wrapper = createComponent();
+      const dayKey = `instructor_email_digests.days.${day}`;
+
+      // When a checkbox is changed, the setter is called with a callback.
+      // We need to capture that callback in order to later verify what it
+      // would have returned.
+      let callback;
+      fakeSetSelectedDays.callsFake(c => {
+        callback = c;
+      });
```

Comment:
I think this is a design flaw.

I used the `StateUpdater` type because at first the state was internal to the `EmailNotificationsPreferences`, and then I realized it had to be a controlled component and moved the state management to `EmailNoitificationsApp`.

This incorrectly conditioned the design of `EmailNotificationsPreferences` props API, and made things too complex.

I see now that `setSelectedDays` should be a simple setter which accepts a `Partial<EmailNotificationsPreferences>`, and then the parent `EmailNoitificationsApp` component, which owns the state updater, can call `setSelectedDays(prev => ({ ...prev, ..updated })`.

I'll do the changes.